Backport from paradise/macros

xeno-by opened the pull request:
Follow-up for https://github.com/scala/scala/pull/2577. Review @paulp @retronym

xeno-by committed:
unifies handling of T's in various analyses of Foo[T]'s

xeno-by committed:
cleans up 82f0925

Updates comments, implements accidentally forgotten IMPLPARAM_TREE,
creates a test to ensure that nothing else is overseen.

paulp commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -431,29 +427,15 @@ trait Definitions extends api.StandardDefinitions {
-      case _                                  => false
-    }
+    // wrapping and unwrapping
+    def dropByName(tp: Type): Type                           = elementExtract(ByNameParamClass, tp).orElse(tp)
```
Let's write this

```
elementExtract(ByNameParamClass, tp) orElse tp
```

It's the most natural place there is to drop punctuation, especially given the english nature of "orElse".

paulp commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -431,29 +427,15 @@ trait Definitions extends api.StandardDefinitions {
+    // wrapping and unwrapping
+    def dropByName(tp: Type): Type                           = elementExtract(ByNameParamClass, tp).orElse(tp)
+    def repeatedToSingle(tp: Type): Type                     = elementExtract(RepeatedParamClass, tp).orElse(tp)
+    def repeatedToSeq(tp: Type): Type                        = elementTransform(RepeatedParamClass, tp)(seqType(_)).orElse(tp)
```
Unless seqType is overloaded or something, this can/should be written

```
elementTransform(RepeatedParamClass, tp)(seqType) orElse tp
```

paulp commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -3700,6 +3701,43 @@ trait Types
+      case NoType => None
+      case tp => Some(tp)
+    }
+  }
```
There's a toOption method on Symbol (NoSymbol == None) which seems a better way, but I wouldn't get too attached to this route, because "any day now" we should have option-less extractors, such that NoType itself can be the failed extraction. So I would prefer not to be adding code of this kind right now.

paulp commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -3700,6 +3701,43 @@ trait Types
+      case tp @ ExistentialType(_, underlying)      => tp.copy(underlying = loop(underlying))
+      case tp @ PolyType(_, resultType)             => tp.copy(resultType = loop(resultType))
+      case tp @ NullaryMethodType(resultType)       => tp.copy(resultType = loop(resultType))
+      case tp                                       => elementTransform(container, tp)(el => appliedType(container, f(el))).orElse(f(tp))
```
This would be crazy expensive in the general case, because it makes no effort whatsoever to avoid creating new objects. Every Annotated/Existential/Poly/Nullary type is copied on every call, even if the transform itself is the identity.

TypeMaps have logic to avoid unnecessary copies; that's much of why they exist. You shouldn't be writing anything which transforms Types which does not leverage TypeMap.

xeno-by replied:
Type maps are deep, not shallow, right? So I'll have to modify them somehow?

paulp replied:
Mmm, I didn't notice this was shallow. The issue remains, but it's less clear you should use TypeMap. OTOH if it wasn't TypeMap it should be some not-yet-existing parent of TypeMap, since it will still overlap heavily with TypeMap.

paulp replied:
Things being shallow vs. deep, intentionally and unintentionally, is such a huge source of bugs that it would be best to think about it without regard for what exists.

xeno-by replied:
I renamed the method to `transparentShallowTransform` to avoid confusion. The name is clumsy, but on the other hand it's unambiguous. Also, such a name is still better than copy & paste. WDYT?

xeno-by committed:
fixes a crash on a degenerate macro definition

Previous version of the MacroImplReference extractor didn't take into
the account the fact that RefTree.qualifier.symbol can be null (and it can
be null if RefTree is an Ident, because then qualifier is an EmptyTree).

This led to NPEs for really weird macro defs that refer to local methods
as their corresponding macro impls. Now I check for this corner case,
and the stuff now longer crashes.

This was wrong; this is how I fixed it; the world is now a better place.

xeno-by commented:
Addressed the comments, also added 04cac3d that fixes a rare bug that I encountered when preparing the slides for ScalaDays.

xeno-by commented:
Ping. The changes are quite small, so maybe you folks could please take a look at them? Would be nice to have relaxed rules for macro implementations in by ScalaDays.

paulp commented:
This is OK with me. I'll merge it today unless retronym sends a minus vote.

xeno-by commented:
Thank you!

retronym commented:
The diffs LGTM, but I think the one-line commit comments ought to be supplemented with "this was wrong; this is how I fixed it; the world is now a better place" elaboration.

xeno-by commented:
Will do

xeno-by commented:
Addressed the recent feedback
